kornia/augmentation/tormentor_minimal/base_augmentation.py

```diff
@@ -0,0 +1,175 @@
+import torch
+from itertools import count
+import types
+
+from functools import wraps
+
+
+class AugmentationFactory(torch.nn.Module):
+    """Generates deterministic augmentations.
+
+    It is a serializable generator that can override default random parameters.
+
+    """
+    def __init__(self, augmentation_cls, **kwargs):
+        super().__init__()
+        assert set(kwargs.keys()) <= set(augmentation_cls.default_params.keys())
+        self.augmentation_params = kwargs
+        self.augmentation_cls = augmentation_cls
+
+    def forward(self, x):
+        if self.training:
+            return self.new()(x)
+        else:
+            return x
+
+    def new(self):
+        return self.augmentation_cls(**self.augmentation_params)
+
+
+class DeterministicImageAugmentation(object):
+    """Deterministic augmentation functor and its factory.
+
+    This class is the base class realising an augmentation.
+    In order to create a new augmentation the forward_sample method and the factory class-function have to be defined.
+    If a constructor is defined it must call super().__init__(**kwargs).
+
+    The **kwargs on the constructor define all needed variables for generating random augmentations.
+    The factory method returns a lambda that instanciates augmentatations.
+    Any parameter about the augmentations randomness should be passed by the factory to the constructor and used inside
+    forward_sample's definition.
+    """
+    _state_names = []
+    _ids = count(0)
+    # since instance seeds are the global seed plus the instance counter, starting from 1000000000 makes a collision
+    # highly improbable. Although no guaranty of uniqueness of instance seed is made.
+    # in case of a multiprocess parallelism, this minimizes chances of collision
+    _global_seed = torch.LongTensor(1).random_(1000000000, 2000000000).item()
+
+    @staticmethod
+    def reset_all_seeds(global_seed=None):
+        SpatialImageAugmentation._ids = count(0)
+        if global_seed is None:
+            global_seed = 0
+        SpatialImageAugmentation._global_seed = global_seed
+
+
+    def __init__(self, id, seed):
+        if id is None:
+            self.id = next(SpatialImageAugmentation._ids)
+        else:
+            self.id = id
+        if seed is None:
+            self.seed = self.id + SpatialImageAugmentation._global_seed
+        else:
+            self.seed = seed
+
+    def __call__(self, tensor_image):
+        device = tensor_image.device
+        with torch.random.fork_rng(devices=(device,)):
+            torch.manual_seed(self.seed)
```

Comment:
For me forking the RNG provides absolute determinism regardless of what the children classes did as long as they stay in a single thread/process don't use foreign random generators eg. numpy.
People writing an augmentation use anything they want torch.rand etc, it also plays well with torch.distributions.
On my machine it takes 780 ns on torch.random.fork_rng(devices=["cuda","cpu"])
This would be an added cost per __call__. In the use case I have in mind this would be acceptable, by batch augmentation makes it even cheaper.

Reply:
basically what I understand here is to decided using `torch.random.fork` vs taking global `torch.Generator`, right ? @shijianjian how's your stuff in terms of memory footprint, @anguelos approach seems quite light-weighted just storing the random seed with just an integer.

Reply:
The main reason for forking the random seed is thread-safety/determinism. In a single thread a cascade of torch.rand, will always be the same for the same seed but if you have multi-thread parallelism you lose determinism. Also careful if you fork, the two forks might end up doing exactly the same random stuff.